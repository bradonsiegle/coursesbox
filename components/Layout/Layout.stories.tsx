import { ComponentStoryObj, ComponentMeta } from "@storybook/react";
import { screen, userEvent } from "@storybook/testing-library";

import { Layout } from "./Layout";

export default {
  title: "Layout/Layout",
  component: Layout,
} as ComponentMeta<typeof Layout>;

export const BasicLayout: ComponentStoryObj<typeof Layout> = {
  play: async ({ args }) => {
    await userEvent.click(screen.getByRole("button"));
  },
  args: {
    children: (
      <>
        <h1>Layout</h1>
        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed euismod,
          nisl nec ultricies lacinia, nisl nisl aliquet nisl, vitae aliquam nisl
          nisl sit amet lorem. Nulla facilisi. Sed euismod, nisl nec ultricies
          lacinia, nisl nisl aliquet nisl, vitae aliquam nisl nisl sit amet
          lorem. Nulla facilisi. Sed euismod, nisl nec ultricies lacinia, nisl
          nisl aliquet nisl, vitae aliquam nisl nisl sit amet lorem. Nulla
          facilisi. Sed euismod, nisl nec ultricies lacinia, nisl nisl aliquet
          nisl, vitae aliquam nisl nisl sit amet lorem. Nulla facilisi. Sed
        </p>
      </>
    ),
  },
};
