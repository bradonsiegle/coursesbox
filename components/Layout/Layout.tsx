import { FC } from 'react';
import styled from '@emotion/styled';

import { Logo } from '@/components/Logo';
