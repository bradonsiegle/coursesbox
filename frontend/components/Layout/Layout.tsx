import Link from "next/link";
import { useSelector, useDispatch } from "react-redux";
import { AppDispatch, RootState } from "@/store";
import { FC, useState, useEffect, useLayoutEffect } from "react";
import { IconButton } from "@/components/IconButton";
import { StyledLink } from "../StyledLink";
import { ThemeProvider } from "@emotion/react";
import { Themes } from "@/styles/themes";
import { login, selectUser } from "@/services/userSlice";

import {
  Wrapper,
  LogoLink,
  StyledLogo,
  MainNav,
  SearchInput,
  Content,
  Footer,
} from "./components";

interface Props {
  children?: React.ReactNode;
}

const useIsomorphicLayoutEffect =
  typeof window !== "undefined" ? useLayoutEffect : useEffect;

export const Layout: FC<Props> = ({ children }) => {
  const { username } = useSelector<RootState, RootState["user"]>(selectUser);

  const [isDark, setIsDark] = useState(true);
  const dispatch = useDispatch<AppDispatch>();

  const toggleDark = () => {
    localStorage.setItem("theme", isDark ? "light" : "dark");
    setIsDark(!isDark);
  };

  useIsomorphicLayoutEffect(() => {
    dispatch(login());

    const theme = localStorage.getItem("theme");
    const themeExistsInStorage = Boolean(theme !== null);

    const isDark = themeExistsInStorage
      ? Boolean(theme === "dark")
      : window.matchMedia("(prefers-color-scheme: dark)").matches;
    setIsDark(isDark);
  }, []);

  const theme = Themes[isDark ? "dark" : "light"];

  return (
    <ThemeProvider theme={theme}>
      <Wrapper>
        <Link href="/" passHref>
          <LogoLink>
            <StyledLogo size={3}>
              <span className="logo_short">CBOX</span>
              <span className="logo_long">CoursesBox</span>
            </StyledLogo>
          </LogoLink>
        </Link>
        <MainNav>
          <Link href="/all" passHref>
            <StyledLink>All</StyledLink>
          </Link>

          <Link href={username ? "/user" : "/login"} passHref>
            <IconButton name={username ? "User" : "Login"} size={1} />
          </Link>

          <IconButton
            name={!isDark ? "Moon" : "Sun"}
            size={1}
            onClick={toggleDark}
          />
        </MainNav>
        <SearchInput icon="Search" placeholder="Search" onChange={() => null} />
        <Content>{children}</Content>
        <Footer>
          <p>Â© 2022 Bradon Siegle. All rights reserved.</p>
        </Footer>
      </Wrapper>
    </ThemeProvider>
  );
};
