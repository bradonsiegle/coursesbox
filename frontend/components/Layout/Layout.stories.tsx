import { ComponentStoryObj, ComponentMeta } from "@storybook/react";
import { screen, userEvent } from "@storybook/testing-library";
import { Provider } from "react-redux";
import { store } from "@/store";

import { Layout } from "./Layout";

export default {
  title: "Layout/Layout",
  component: Layout,
} as ComponentMeta<typeof Layout>;

export const BasicLayout: ComponentStoryObj<typeof Layout> = {
  args: {
    children: (
      <>
        <h1>Layout</h1>
        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed euismod,
          nisl nec ultricies lacinia, nisl nisl aliquet nisl, vitae aliquam nisl
          nisl sit amet lorem. Nulla facilisi. Sed euismod, nisl nec ultricies
          lacinia, nisl nisl aliquet nisl, vitae aliquam nisl nisl sit amet
          lorem. Nulla facilisi. Sed euismod, nisl nec ultricies lacinia, nisl
          nisl aliquet nisl, vitae aliquam nisl nisl sit amet lorem. Nulla
          facilisi. Sed euismod, nisl nec ultricies lacinia, nisl nisl aliquet
          nisl, vitae aliquam nisl nisl sit amet lorem. Nulla facilisi. Sed
        </p>
      </>
    ),
  },
  decorators: [
    (Story) => (
      <Provider store={store}>
        <Story />
      </Provider>
    ),
  ],
};
