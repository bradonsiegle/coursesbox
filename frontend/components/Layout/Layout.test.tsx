import { render, screen, act } from "@/test-utils";
import userEvent from "@testing-library/user-event";
import { Layout } from "./Layout";

describe("Layout test cases", () => {
  beforeAll(() => {
    Object.defineProperty(window, "matchMedia", {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });

  const child = (
    <>
      <h1>Main Article</h1>
      <p>Some text</p>
    </>
  );

  it("should render the layout", () => {
    const { container } = render(<Layout>{child}</Layout>);
    expect(container).toMatchSnapshot();
  });
  it("should toggle theme", async () => {
    localStorage.setItem("theme", "light");
    (window.matchMedia as jest.Mock).mockReturnValue({ matches: true });

    render(<Layout>{child}</Layout>);

    const themeToggler = screen.getByRole("button", { name: "Moon" });
    expect(themeToggler).toBeInTheDocument();

    await act(async () => {
      userEvent.click(themeToggler);
    });

    expect(localStorage.getItem("theme")).toBe("dark");
    expect(screen.getByRole("button", { name: "Sun" })).toBeInTheDocument();
  });
});
