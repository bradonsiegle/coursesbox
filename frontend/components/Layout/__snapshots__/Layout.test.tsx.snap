// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout test cases Render check 1`] = `
<DocumentFragment>
  <div
    class="css-1stdm9q"
  >
    <a
      class="css-1m1nve0"
      href="/"
    >
      <header
        class="css-w7ne20"
        size="1.4"
      >
        <h2>
          RecipeHub
        </h2>
      </header>
    </a>
    <nav
      class="css-1sstk05"
    >
      <button
        class="css-dujfmn"
        title="BiBookmarkHeart"
      >
        <svg
          aria-label="BiBookmarkHeart"
          class="css-8mpkb1"
          fill="currentColor"
          height="1em"
          href="/likes"
          role="img"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M20 22V4c0-1.103-.897-2-2-2H6c-1.103 0-2 .897-2 2v18l8-4.572L20 22zM6 10V4h12v14.553l-6-3.428-6 3.428V10z"
          />
          <path
            d="M15.409 9.512c.361-.372.585-.888.585-1.456s-.223-1.083-.585-1.456a1.962 1.962 0 0 0-1.412-.603S13.001 5.994 12 7.026c-1.001-1.032-1.997-1.029-1.997-1.029-.552 0-1.051.23-1.412.603-.362.373-.585.887-.585 1.456s.223 1.084.585 1.456L12 13.203l3.409-3.691z"
          />
        </svg>
      </button>
      <button
        class="css-dujfmn"
        title="Login"
      >
        <svg
          aria-hidden="true"
          aria-label="Login"
          class="css-8mpkb1"
          data-icon="login"
          fill="currentColor"
          focusable="false"
          height="1rem"
          href="/login"
          role="img"
          viewBox="64 64 896 896"
          width="1rem"
        >
          <defs>
            <style />
          </defs>
          <path
            d="M521.7 82c-152.5-.4-286.7 78.5-363.4 197.7-3.4 5.3.4 12.3 6.7 12.3h70.3c4.8 0 9.3-2.1 12.3-5.8 7-8.5 14.5-16.7 22.4-24.5 32.6-32.5 70.5-58.1 112.7-75.9 43.6-18.4 90-27.8 137.9-27.8 47.9 0 94.3 9.3 137.9 27.8 42.2 17.8 80.1 43.4 112.7 75.9 32.6 32.5 58.1 70.4 76 112.5C865.7 417.8 875 464.1 875 512c0 47.9-9.4 94.2-27.8 137.8-17.8 42.1-43.4 80-76 112.5s-70.5 58.1-112.7 75.9A352.8 352.8 0 01520.6 866c-47.9 0-94.3-9.4-137.9-27.8A353.84 353.84 0 01270 762.3c-7.9-7.9-15.3-16.1-22.4-24.5-3-3.7-7.6-5.8-12.3-5.8H165c-6.3 0-10.2 7-6.7 12.3C234.9 863.2 368.5 942 520.6 942c236.2 0 428-190.1 430.4-425.6C953.4 277.1 761.3 82.6 521.7 82zM395.02 624v-76h-314c-4.4 0-8-3.6-8-8v-56c0-4.4 3.6-8 8-8h314v-76c0-6.7 7.8-10.5 13-6.3l141.9 112a8 8 0 010 12.6l-141.9 112c-5.2 4.1-13 .4-13-6.3z"
          />
        </svg>
      </button>
      <button
        class="css-dujfmn"
        title="Moon"
      >
        <svg
          aria-label="Moon"
          class="css-8mpkb1"
          fill="currentColor"
          height="1rem"
          role="img"
          viewBox="0 0 24 24"
          width="1rem"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M10 6a8 8 0 0011.955 6.956C21.474 18.03 17.2 22 12 22 6.477 22 2 17.523 2 12c0-5.2 3.97-9.474 9.044-9.955A7.963 7.963 0 0010 6zm-6 6a8 8 0 008 8 8.006 8.006 0 006.957-4.045c-.316.03-.636.045-.957.045-5.523 0-10-4.477-10-10 0-.321.015-.64.045-.957A8.006 8.006 0 004 12zm14.164-9.709L19 2.5v1l-.836.209a2 2 0 00-1.455 1.455L16.5 6h-1l-.209-.836a2 2 0 00-1.455-1.455L13 3.5v-1l.836-.209A2 2 0 0015.29.836L15.5 0h1l.209.836a2 2 0 001.455 1.455zm5 5L24 7.5v1l-.836.209a2 2 0 00-1.455 1.455L21.5 11h-1l-.209-.836a2 2 0 00-1.455-1.455L18 8.5v-1l.836-.209a2 2 0 001.455-1.455L20.5 5h1l.209.836a2 2 0 001.455 1.455z"
          />
        </svg>
      </button>
    </nav>
    <label
      class="css-153eln6"
      height="7"
      width="20"
    >
      <span
        class="css-1grri41"
      />
      <div
        class="css-52ovgc"
      >
        <input
          class="css-9em8xk"
          placeholder="Find a recipe"
          value=""
        />
        <svg
          aria-hidden="true"
          aria-label="Search"
          class="css-153tpqd"
          data-icon="search"
          fill="currentColor"
          focusable="false"
          height="1.6rem"
          role="img"
          viewBox="64 64 896 896"
          width="1.6rem"
        >
          <path
            d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
          />
        </svg>
      </div>
      <span
        class="css-1reg724"
        role="alert"
      />
    </label>
    <main
      class="css-tee3o6"
    >
      <h1>
        Main article area
      </h1>
      <p>
        In this layout, we display the areas in source order for any screen less that 500 pixels wide. We go to a two column layout, and then to a three column layout by redefining the grid, and the placement of items on the grid.
      </p>
    </main>
    <footer
      class="css-1y0dtms"
    >
      <p>
        Â© 2022 Bradon Siegle. All rights reserved.
      </p>
    </footer>
  </div>
</DocumentFragment>
`;
